olive/resource_path.py

```diff
@@ -0,0 +1,387 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import inspect
+import logging
+import shutil
+import tempfile
+from abc import abstractmethod
+from enum import Enum
+from pathlib import Path
+from typing import Any, Dict, Type, Union
+
+from pydantic import Field, validator
+
+from olive.azureml.azureml_client import AzureMLClientConfig
+from olive.common.config_utils import ConfigBase, validate_config
+from olive.common.utils import retry_func
+
+logger = logging.getLogger(__name__)
+
+
+class ResourceType(str, Enum):
+    LocalFile = "file"
+    LocalFolder = "folder"
+    StringName = "string_name"
+    AzureMLModel = "azureml_model"
+    AzureMLDatastore = "azureml_datastore"
+    AzureMLJobOutput = "azureml_job_output"
+
+    def __str__(self) -> str:
+        return self.value
+
+
+LOCAL_RESOURCE_TYPES = [ResourceType.LocalFile, ResourceType.LocalFolder, ResourceType.StringName]
+AZUREML_RESOURCE_TYPES = [ResourceType.AzureMLModel, ResourceType.AzureMLDatastore, ResourceType.AzureMLJobOutput]
+
+_resource_config_registry: Dict[ResourceType, Type[ConfigBase]] = {}
+
+
+class ResourceConfig(ConfigBase):
```

Comment:
could we unify the hf_config by using the ResourceConfig?

Reply:
`hf_config` is not only used to load models but also hold information about the hf configuration of the original model. It is also inherited by output models. So, it is not the same as resource path. 

You can see here that the model_path can be used to load the model even when there is hf config https://github.com/microsoft/Olive/blob/main/olive/model.py#L568/ 

I think we can make changes in a separate PR if we want to rethink this part.

Reply:
We can reconsider whether hf_config could be aligned with the resource path concept in future.